File: leetcode/algorithms/203. 移除链表元素（没一遍过）/cpp/Solution1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        ListNode* cur = dummy;
        while(cur&&cur->next){
            if(cur->next->val==val){
                ListNode* tmp = cur->next;
                cur->next = cur->next->next;
                delete tmp;
            }else{
                cur = cur->next;
            }
        }
        return dummy->next;
    }
};

File: leetcode/algorithms/203. 移除链表元素（没一遍过）/java/Solution1.java
class Solution1 {
  public ListNode removeElements(ListNode head, int val) {
      // 若一开始那一部分都是需要被删除的元素
      while(head != null && head.val == val){
          ListNode delNode = head;
          head = head.next;
          //方便java垃圾回收
          delNode.next = null;
      }
      // 若全部都是要被删除的节点
      if(head == null)
          return head;
      ListNode prev = head;
      while(prev.next != null){
          if(prev.next.val == val) {
              ListNode delNode = prev.next;
              prev.next = delNode.next;
              delNode.next = null;
          }
          else{
              prev = prev.next;
          }
      }
      return head;
  }
}
// or
/*
class Solution1 {
    public ListNode removeElements(ListNode head, int val) {
        while(head != null && head.val == val)
            head = head.next;
        if(head == null)
            return head;
        ListNode prev = head;
        while(prev.next != null){
            if(prev.next.val == val)
                prev.next = prev.next.next;
            else
                prev = prev.next;
        }
        return head;
    }
}
*/